test/addons-napi/test_promise/test_promise.c

```diff
@@ -0,0 +1,65 @@
+#include <node_api.h>
+#include "../common.h"
+
+napi_ref deferred_ref = NULL;
+
+napi_value createPromise(napi_env env, napi_callback_info info) {
+  napi_value promise, deferred, in_one_call;
+  bool one_call;
+  size_t argc = 1;
+
+  NAPI_CALL(env, napi_get_cb_info(env, info, &argc, &in_one_call, NULL, NULL));
+  NAPI_CALL(env, napi_get_value_bool(env, in_one_call, &one_call));
+
+  if (deferred_ref != NULL) {
+    NAPI_CALL(env, napi_delete_reference(env, deferred_ref));
+  }
+
+  if (one_call) {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, &promise));
+  } else {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, NULL));
+    NAPI_CALL(env, napi_get_promise_from_deferred(env, deferred, &promise));
+  }
+  NAPI_CALL(env, napi_create_reference(env, deferred, 1, &deferred_ref));
+
+  return promise;
+}
```

Comment:
@TimothyGu the fact that a `v8::Deferred === v8::Promise` seems like a shortcoming of V8. The only thing I can think of doing on the N-API side to enforce the distinction between a deferred and a promise is somewhat convoluted:

* To cover creation:
  1. Declare a JS class "Deferred" and give it an internal field. This would currently have to be done once per N-API module, because we don't have proper per-isolate storage. This means that passing a "Deferred" between N-API modules would not work - which is perhaps not a bad thing.
  2. Store a persistent reference to an instance of `v8::Resolver` in the internal field of an instance of `Deferred`.
  3. Allow the instance of `Deferred` to be passed into JS.
* To cover usage:
  1. Given a `napi_value` that claims to be a `Deferred`, test it with `napi_instanceof()` against the `Deferred` constructor.
  2. If it passes, then retrieve the internal field and obtain a local reference to the `Resolver` from the persistent reference.

The only alternative I can currently think of is documentation. Put bluntly, "Yes, you can pass a Deferred into JS, yes it may even behave like a promise, but we're telling you not to do that, because it's not guaranteed to behave like a promise if we ever swap out the underlying engine or if V8 decides to introduce a distinction, and we will make no efforts to support that use case given that you can easily retrieve a bona fide JavaScript promise from a Deferred."